File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
n, m = map(int, input().split())
l = list(map(int, input().split()))
l.sort()
left = 0;right = l[-1]
sum = 0
while left <= right:
    temp_sum = sum
    sum = 0
    middle = (left + right)//2
    for i in l: sum += ((i - middle) if i > middle else 0)
    if sum < m: right = middle - 1
    elif sum >= m : left = middle + 1
print(right)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    int n;
    long long m;
    std::cin >> n >> m;

    
    std::vector<int> v(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> v[i];
    }

    std::sort(v.begin(), v.end());

    int left = 0;
    int right = v[n - 1];

    while (left <= right) {
        int middle = (left + right) / 2;
        long long sum = 0;

        for (int i = 0; i < n; ++i) {
            if (v[i] > middle) {
                sum += v[i] - middle;
            }
        }

        if (sum < m) {
            right = middle - 1;
        }
        else {
            left = middle + 1;
        }
    }

    std::cout << right;
}